pkg/controllers/tagging/tagging_controller.go

```diff
@@ -0,0 +1,289 @@
+// It periodically check for Node events (creating/deleting) to apply appropriate
+// tags to resources.
+type TaggingController struct {
+	nodeInformer coreinformers.NodeInformer
+	kubeClient   clientset.Interface
+	cloud        *awsv1.Cloud
+	workqueue    workqueue.RateLimitingInterface
+
+	// Value controlling TaggingController monitoring period, i.e. how often does TaggingController
+	// check node list. This value should be lower than nodeMonitorGracePeriod
+	// set in controller-manager
+	nodeMonitorPeriod time.Duration
+
+	// Representing the user input for tags
+	tags map[string]string
+
+	// Representing the resources to tag
+	resources []string
+}
+
+// NewTaggingController creates a NewTaggingController object
+func NewTaggingController(
+	nodeInformer coreinformers.NodeInformer,
+	kubeClient clientset.Interface,
+	cloud cloudprovider.Interface,
+	nodeMonitorPeriod time.Duration,
+	tags map[string]string,
+	resources []string) (*TaggingController, error) {
+
+	awsCloud, ok := cloud.(*awsv1.Cloud)
+	if !ok {
+		err := fmt.Errorf("tagging controller does not support %v provider", cloud.ProviderName())
+		return nil, err
+	}
+
+	tc := &TaggingController{
+		nodeInformer:      nodeInformer,
+		kubeClient:        kubeClient,
+		cloud:             awsCloud,
+		nodeMonitorPeriod: nodeMonitorPeriod,
+		tags:              tags,
+		resources:         resources,
+		workqueue:         workqueue.NewNamedRateLimitingQueue(workqueue.DefaultControllerRateLimiter(), "Tagging"),
+	}
+
+	// Use shared informer to listen to add/update/delete of nodes. Note that any nodes
+	// that exist before tagging controller starts will show up in the update method
+	tc.nodeInformer.Informer().AddEventHandler(cache.ResourceEventHandlerFuncs{
+		AddFunc:    func(obj interface{}) { tc.enqueueNode(obj, true) },
+		UpdateFunc: func(oldObj, newObj interface{}) { tc.enqueueNode(newObj, true) },
+		DeleteFunc: func(obj interface{}) { tc.enqueueNode(obj, false) },
+	})
+
+	return tc, nil
+}
+
+// Run will start the controller to tag resources attached to the cluster
+// and untag resources detached from the cluster.
+func (tc *TaggingController) Run(stopCh <-chan struct{}) {
+	defer utilruntime.HandleCrash()
+	defer tc.workqueue.ShutDown()
+
+	klog.Infof("Starting the tagging controller")
+	go wait.Until(tc.work, tc.nodeMonitorPeriod, stopCh)
+
+	<-stopCh
+}
+
+// work is a long-running function that continuously
+// call process() for each message on the workqueue
+func (tc *TaggingController) work() {
+	for tc.Process() {
+	}
+}
+
+// Process reads each message in the queue and performs either
+// tag or untag function on the Node object
+func (tc *TaggingController) Process() bool {
+	obj, shutdown := tc.workqueue.Get()
+	if shutdown {
+		return false
+	}
+
+	klog.Infof("Starting to process %v", obj)
+
+	err := func(obj interface{}) error {
+		defer tc.workqueue.Done(obj)
+
+		var key string
+		var ok bool
+		if key, ok = obj.(string); !ok {
```

Comment:
In our case, `obj` is a node object, right? Are we able to successfully convert it to `string` here?

Reply:
`obj` is not a node object in this case because we transformed that into a key in [here](https://github.com/kubernetes/cloud-provider-aws/pull/308/files#diff-ac633e66d7cfdfa6934f05558a33311d493f704ab69e5443c4ad046a47ae1266R265). (BTW, I'm not opposed to putting the whole object string in as key either, but I don't know if that's gonna work because this is how the node_controller enqueue key and get node from key, so we need to do some experimenting to find out, which I don't see the benefits of yet)